server/app/uk/gov/ons/addressIndex/server/modules/AddressIndexActions.scala

```diff
@@ -0,0 +1,114 @@
+    def input: T
+  }
+  case class UprnQueryInput(override val input: String) extends QueryInput[String]
+  case class AddressQueryInput(override val input: Seq[CrfTokenResult]) extends QueryInput[Seq[CrfTokenResult]]
+
+  /**
+    * @param tokens
+    * @return
+    */
+  def pafSearch(tokens: QueryInput[Seq[CrfTokenResult]]): Future[AddressBySearchResponseContainer] = {
+    esRepo queryPafAddresses tokens.input map { case PostcodeAddressFileAddresses(addresses, maxScore) =>
+      searchContainerTemplate(
+        tokens = tokens.input,
+        addresses = addresses map(AddressResponseAddress fromPafAddress maxScore),
+        total = addresses.size
+      )
+    }
+  }
+
+  /**
+    * @param tokens
+    * @return
+    */
+  def nagSearch(tokens: QueryInput[Seq[CrfTokenResult]]): Future[AddressBySearchResponseContainer] = {
+    esRepo queryNagAddresses tokens.input map { case NationalAddressGazetteerAddresses(addresses, maxScore) =>
+      searchContainerTemplate(
+        tokens = tokens.input,
+        addresses = addresses map(AddressResponseAddress fromNagAddress maxScore),
+        total = addresses.size
+      )
+    }
+  }
+
+  /**
+    * @param uprn
+    * @return
+    */
+  def uprnPafSearch(uprn: QueryInput[String]): Future[AddressByUprnResponseContainer] = {
+    esRepo queryPafUprn uprn.input map {
+      _.map { address =>
+        searchUprnContainerTemplate(
+          Some(AddressResponseAddress fromPafAddress address)
+        )
+      } getOrElse NoAddressFoundUprn
+    }
+  }
+
+  /**
+    * @param uprn
+    * @return
+    */
+  def uprnNagSearch(uprn: QueryInput[String]): Future[AddressByUprnResponseContainer] = {
+    esRepo queryNagUprn uprn.input map {
+      _.map { address =>
+        searchUprnContainerTemplate(
+          Some(AddressResponseAddress fromNagAddress address)
+        )
+      } getOrElse NoAddressFoundUprn
+    }
+  }
+
+  /**
+    * This is a PAF or NAG switch helper which can be used for creating a Future[Ok[Json]]
+    *
+    * @param formatStr the input format String
+    * @param inputForPafFn the input for pafFn
+    * @param pafFn the function which will be called if the formatStr resolves to `PostcodeAddressFile`
+    * @param inputForNagFn  the input for nagFn
+    * @param nagFn the function which will be called if the formatStr resolves to `BritishStandard7666`
+    * @tparam T the return type of the object which will be "PlayJson'd"
+    * @tparam QueryInputType the input type for the Query
+    * @return If None, the formatStr failed to resole.
+    *         If Some, the appropriate object for the given function and resolved format.
+    */
+  def formatQuery[T, QueryInputType](
+    formatStr: String,
+    inputForPafFn: QueryInput[QueryInputType],
+    pafFn: QueryInput[QueryInputType] => Future[T],
+    inputForNagFn: QueryInput[QueryInputType],
+    nagFn: QueryInput[QueryInputType] => Future[T]
+  )(implicit writes: Writes[T]): Option[Future[Result]] = {
+    (
```

Comment:
I'm, maybe, missing something, but will this code work if you put `map` directly after `}` ? (I know that this isn't the case for for comprehensions, but it's not the case here)